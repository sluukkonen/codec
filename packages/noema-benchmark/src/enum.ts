import * as n from "noema"
import * as z from "zod"
import { Benchmark } from "./index.js"

const suite: Benchmark<string> = {
  name: "enum",
  data: "foo",
  codecs: { noema: n.enum("foo", "bar"), zod: z.enum(["foo", "bar"]) },
}

export default suite
