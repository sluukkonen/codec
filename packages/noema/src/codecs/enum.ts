import { createSimpleCodec, SimpleCodec } from "../Codec.js"
import { Literal } from "./literal.js"
import { EnumMetadata } from "../Metadata.js"

type EnumCodec<T extends Literal> = SimpleCodec<T, EnumMetadata<T>>

function enumCodec<T extends Literal>(...members: T[]): EnumCodec<T> {
  const set = new Set(members)
  return createSimpleCodec(
    (val, ctx) =>
      set.has(val as T)
        ? ctx.success(val as T)
        : ctx.failure({
            code: "invalid_enum",
            path: ctx.path,
            members,
          }),
    { tag: "enum", simple: true, members }
  )
}

export { enumCodec as enum }
