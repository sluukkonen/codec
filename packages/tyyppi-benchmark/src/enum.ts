import * as t from "tyyppi"
import * as z from "zod"
import { Benchmark } from "./index.js"

const suite: Benchmark<string> = {
  name: "enum",
  data: "foo",
  codecs: { tyyppi: t.enum("foo", "bar"), zod: z.enum(["foo", "bar"]) },
}

export default suite
