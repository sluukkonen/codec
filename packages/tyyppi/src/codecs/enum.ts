import { createSimpleCodec, SimpleCodec, SimpleMetadata } from "../Codec.js"
import { invalidEnum, InvalidEnum } from "../DecodeError.js"
import { failure, success } from "../Result.js"
import { Primitive } from "../types.js"

interface EnumMetadata<T extends Primitive> extends SimpleMetadata {
  readonly tag: "enum"
  readonly members: readonly T[]
}

export type EnumCodec<T extends Primitive> = SimpleCodec<
  T,
  InvalidEnum<T>,
  EnumMetadata<T>
>

function enumCodec<T extends Primitive>(
  ...members: readonly T[]
): EnumCodec<T> {
  const set = new Set(members)
  return createSimpleCodec(
    (val) =>
      set.has(val as T)
        ? success(val as T)
        : failure(invalidEnum(val, members)),
    { tag: "enum", simple: true, members },
  )
}

export { enumCodec, enumCodec as enum }
