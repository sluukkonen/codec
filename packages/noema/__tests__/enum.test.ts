import { expectParseFailure, expectParseSuccess } from "./helpers.js"
import { enum as enumCodec } from "../src/codecs/enum.js"

describe("enum", () => {
  test("should parse valid members", () => {
    expectParseSuccess(enumCodec(1, 2), 1)
    expectParseSuccess(enumCodec(1, 2), 2)
  })

  test("should reject other values", () => {
    expectParseFailure(enumCodec(1, 2), 3)
  })
})
