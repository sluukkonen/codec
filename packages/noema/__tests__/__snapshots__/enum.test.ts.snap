// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enum should reject other values 1`] = `
[
  {
    "code": "invalid_enum",
    "members": [
      1,
      2,
    ],
    "path": "",
  },
]
`;
